dkg/sync/client.go

```diff
@@ -16,84 +16,159 @@
 }
 
-// Shutdown sends a shutdown message to the peer indicating it has successfully completed.
+// Shutdown sends a shutdown message to the server indicating it has successfully completed.
 // It closes the connection and returns after receiving the subsequent MsgSyncResponse.
 // It may only be called after AwaitConnected.
 func (*Client) Shutdown() error {
 	return nil
 }
 
+func sendHashSignature(ctx context.Context, hashSig []byte, s network.Stream) result {
+	before := time.Now()
+	msg := &pb.MsgSync{
+		Timestamp:     timestamppb.Now(),
+		HashSignature: hashSig,
+		Shutdown:      false,
+	}
+
+	b, err := proto.Marshal(msg)
+	if err != nil {
+		log.Error(ctx, "Marshal msg", err)
+		return result{error: err.Error()}
+	}
+
+	if _, err = s.Write(b); err != nil {
+		log.Error(ctx, "Write msg to stream", err)
+		return result{error: err.Error()}
+	}
+
+	buf := bufio.NewReader(s)
+	rb := make([]byte, MsgSize)
+	// n is the number of bytes read from buffer, if n < MsgSize the other bytes will be 0
+	n, err := buf.Read(rb)
+	if err != nil {
+		log.Error(ctx, "Read server response from stream", err)
+		return result{error: err.Error()}
+	}
+
+	// Number of bytes that are read are the most important
+	rb = rb[:n]
+
+	resp := new(pb.MsgSyncResponse)
+	if err = proto.Unmarshal(rb, resp); err != nil {
+		log.Error(ctx, "Unmarshal server response", err)
+		return result{error: err.Error()}
+	}
+
+	log.Debug(ctx, "Server response", z.Any("response", resp.SyncTimestamp))
+
+	return result{
+		rtt:       time.Since(before),
+		timestamp: resp.SyncTimestamp.String(),
+		error:     resp.Error,
+	}
+}
+
 // NewClient starts a goroutine that establishes a long lived connection to a p2p server and returns a new Client instance.
-func NewClient(ctx context.Context, tcpNode host.Host, server p2p.Peer, hashSig []byte, onFailure func(), ch chan *pb.MsgSyncResponse) Client {
-	go func() {
-		s, err := tcpNode.NewStream(ctx, server.ID, syncProtoID)
-		if err != nil {
-			log.Error(ctx, "Open new stream with server", err)
-		}
-		defer s.Close()
+// TODO(dhruv): call onFailure on permanent failure.
+func NewClient(ctx context.Context, tcpNode host.Host, server p2p.Peer, hashSig []byte, onFailure func()) Client {
+	s, err := tcpNode.NewStream(ctx, server.ID, syncProtoID)
+	if err != nil {
+		log.Error(ctx, "Open new stream with server", err)
+		ch := make(chan result, 1)
+		ch <- result{error: err.Error()}
+		close(ch)
 
-		msg := &pb.MsgSync{
-			Timestamp:     timestamppb.Now(),
-			HashSignature: hashSig,
-			Shutdown:      false,
+		return Client{
+			ctx:       ctx,
+			onFailure: onFailure,
+			tcpNode:   tcpNode,
+			server:    server,
+			results:   ch,
 		}
+	}
 
-		b, err := proto.Marshal(msg)
-		if err != nil {
-			log.Error(ctx, "Marshal msg", err)
-		}
+	ctx, cancel := context.WithCancel(ctx)
 
-		if _, err = s.Write(b); err != nil {
-			log.Error(ctx, "Write msg to stream", err)
-		}
+	out := make(chan result)
+	go func() {
+		defer close(out)
```

Comment:
discussion: should this statement be outside of the goroutine?🤔

Reply:
nope, this go routine is responsible of writing into this channel so thats why it needs to close the channel once context is called or ctx.Done is closed